vault/core.go

```diff
@@ -1249,110 +1263,166 @@ func (c *Core) unsealPart(ctx context.Context, seal Seal, key []byte, useRecover
-			}
+		// In this case we have to ensure that the recovery information was
+		// set properly.
+		if c.migrationInfo.recoveryKey == nil {
+			return errors.New("did not get expected recovery information to set new seal during migration")
+		}
 
-			// Return the new key so it can be used to unlock the barrier
-			masterKey = newMasterKey
+		// We have recovery keys; we're going to use them as the new
+		// shamir KeK.
+		err := c.seal.GetAccess().Wrapper.(*aeadwrapper.Wrapper).SetAESGCMKeyBytes(c.migrationInfo.recoveryKey)
+		if err != nil {
+			return errwrap.Wrapf("failed to set master key in seal: {{err}}", err)
+		}
 
-		default:
-			return nil, errors.New("unhandled migration case (shamir to shamir)")
+		if err := c.seal.SetStoredKeys(ctx, [][]byte{c.migrationInfo.masterKey}); err != nil {
+			return errwrap.Wrapf("error setting new barrier key information during migrate: {{err}}", err)
 		}
 
-		// At this point we've swapped things around and need to ensure we
-		// don't migrate again
-		c.migrationSeal = nil
-		atomic.StoreUint32(c.sealMigrated, 1)
+	case c.seal.RecoveryKeySupported():
+		// Migration is happening from shamir -> auto. In this case use the shamir
+		// combined key that was used to store the master key as the new recovery key.
+		if err := c.seal.SetRecoveryKey(ctx, c.migrationInfo.shamirCombinedKey); err != nil {
+			return errwrap.Wrapf("error setting new recovery key information: {{err}}", err)
+		}
 
-		// Ensure we populate the new values
-		bc, err := c.seal.BarrierConfig(ctx)
+		// Generate a new master key
+		newMasterKey, err := c.barrier.GenerateKey(c.secureRandomReader)
 		if err != nil {
-			return nil, errwrap.Wrapf("error fetching barrier config after migration: {{err}}", err)
+			return errwrap.Wrapf("error generating new master key: {{err}}", err)
 		}
-		if err := c.seal.SetBarrierConfig(ctx, bc); err != nil {
-			return nil, errwrap.Wrapf("error storing barrier config after migration: {{err}}", err)
+
+		// Rekey the barrier
+		if err := c.barrier.Rekey(ctx, newMasterKey); err != nil {
+			return errwrap.Wrapf("error rekeying barrier during migration: {{err}}", err)
 		}
 
-		if c.seal.RecoveryKeySupported() {
-			rc, err := c.seal.RecoveryConfig(ctx)
-			if err != nil {
-				return nil, errwrap.Wrapf("error fetching recovery config after migration: {{err}}", err)
-			}
-			if err := c.seal.SetRecoveryConfig(ctx, rc); err != nil {
-				return nil, errwrap.Wrapf("error storing recovery config after migration: {{err}}", err)
-			}
+		// Store the new master key
+		if err := c.seal.SetStoredKeys(ctx, [][]byte{newMasterKey}); err != nil {
+			return errwrap.Wrapf("error storing new master key: {{err}}", err)
 		}
+
+	default:
+		return errors.New("unhandled migration case (shamir to shamir)")
 	}
 
-	return masterKey, nil
+DONE:
+
+	// At this point we've swapped things around and need to ensure we
+	// don't migrate again
+	c.migrationInfo = nil
```

Comment:
Maybe call `memzero` on the key fields before setting it to nil to do a best-effort overwrite on the bytes.

Reply:
Vishal and I talked about the memzeros in previous comments.  He intimated that we weren't so concerned anymore about making this weak effort to avoid copies of the master key floating around.  I'm planning to do a followup PR after this is merged to try to get rid of the existing memzeros, to avoid confusion (do I need to copy this []byte before calling this method?) and promote consistency.